WIP: Post Editor: Revisions sidebar

bperson opened the pull request:
**WIP for the editor's Post Revisions UI.**

- [x] "child" sidebar support
- [x] listing revisions (still need to figure out how to get the author name instead of his ID)
- [x] view revision (tap on one of the revisions in the sidebar)
- [x] load revision in editor
- [x] raw HTML diff

### Current issues
- [x] first load: auto-select the first revision when data is available (instead of at mount
- [x] author ID but no author name
- [x] mobile: icon change in the top toolbar to the "go back in time" icon
- [x] mobile: sidebar header should be visible (and cover the editor's action bar) to make it possible to dismiss the revisions child sidebar.
- [ ] placeholders in the revisions list / diff viewer while loading revisions
- [x] If you're scrolled in the editor, open the revisions sidebar and close it, you will loose your scroll position. (this scroll issue creates misplacement issues with TinyMCE's toolbar)
- [ ] when loading a revision in the editor, show a popup to make sure the user is fine with "loosing" content if anything's not saved

### Open questions
- When loading raw content into TinyMCE in html mode, some tags are visible (`<p>`, `<br />`). A double-switch -> "Visual" -> "HTML" will clear those tags. Is this an issue? **still investigating**

### Further down the line (in subsequent PRs)
- updated "status" sidebar accordion based on @folletto's design
- support for pages / custom post types
- "rendered" HTML diff (not sure)
- display word count in the diff viewer and "update" it to the number of word in the revision currently being displayed

### Screenshot
![screen shot 2017-04-27 at 18 52 47](https://cloud.githubusercontent.com/assets/1145270/25494593/6f6adb20-2b7a-11e7-8327-4045faa5c4ca.png)

a8ci18n commented on client/post-editor/editor-revisions-list/header.jsx:
```diff
@@ -0,0 +1,32 @@
+				onClick={ restoreRevision }
+				compact={ true }
+			>
+				{ translate( 'Restore this revision' ) }
```
Hi! I've found a possible matching string that has already been translated **74** times:
`translate( 'Restore This Revision' )` *ES Score: 17*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

bperson commented:
@folletto / @mtias I don't think it's worth doing a "code" review at this point (considering how big it is, I'll probably split it into multiple smaller independent PRs once it's a bit more stabilised)

I'm more interested in information / pointers / ideas for some of the "open questions".

@folletto feel free to point out anything that I might have missed from the sketches ;)

folletto commented:
I'll wait a full review once it's complete, so far it looks good. The important bit now is to always show a revision when it opens, as you mentioned already above. :)

There are also small details, I'll point out even if I'm aware it's still in progress:

1. The hover state (that wasn't specified in the designs) should be the same as the accordions (blue + light background)
2. "Post Settings" should be linked as it is now, but not change color. Blue hover color.
3. The footer should stay, and I'd probably add a 10px margin at the bottom to make clear the list ends there.
4. The word counter in the bottom-right corner isn't changing. Would be ideal if it updates, but we can start just by removing it for now and do in a later PR.

---

> When loading raw content into TinyMCE in html mode, some tags are visible. A double-switch -> "Visual" -> "HTML" will clear those tags. Is this an issue?

Ideally it shouldn't happen. It's probably due to the "magic" WordPress does behind the scenes to auto-p (`wpautop`) the content...

> popup making sure the user is fine with "loosing" content not yet saved?

Nice catch. Yes. :)

---

Note: even if it's in progress, try to add screenshot of the latest state of the PR when you update. ;)

bperson commented:
> The hover state (that wasn't specified in the designs) should be the same as the accordions (blue + light background)

fixed
> "Post Settings" should be linked as it is now, but not change color. Blue hover color.

fixed

> The word counter in the bottom-right corner isn't changing. Would be ideal if it updates, but we can start just by removing it for now and do in a later PR.

hidden, auto-update has been added to my "further down the line" list :) 

> The footer should stay, and I'd probably add a 10px margin at the bottom to make clear the list ends there.

I initially kept it there, but with "move to trash" I figured it could get confusing: are we moving the post to trash? just the revision currently selected?. That said, there is probably some value in having the help button always visible / accessible?

> Other than the author name... the word count isn't there yet, right?

yep there is support to display the word count diffs but since I'm not computing any diff yet, there is no "data" -> it's just a 10/10 placeholder.

> Ideally it shouldn't happen. It's probably due to the "magic" WordPress does behind the scenes to auto-p (`wpautop`) the content...

will look into that, thanks :)

> Note: even if it's in progress, try to add screenshot of the latest state of the PR when you update. ;) 

done 👌 

Thanks @folletto for this initial review :)

folletto commented:
> I initially kept it there, but with "move to trash" I figured it could get confusing: are we moving the post to trash? just the revision currently selected?. That said, there is probably some value in having the help button always visible / accessible?

If you notice, the footer is meant to be present, but without "Move to trash", so that should answer the question.  :D 
If that's too complicated, then we can keep it even if I agree it's less ideal.

mtias commented on client/post-editor/edit-post-status/index.jsx:
```diff
@@ -41,6 +41,7 @@ class EditPostStatus extends Component {
 		postDate: PropTypes.string,
 		onPrivatePublish: PropTypes.func,
 		status: PropTypes.string,
+		toggleChildSidebar: PropTypes.func,
```
Is `childSidebar` clear enough for someone reading the code?

mtias commented on client/post-editor/editor-diff-viewer/index.jsx:
```diff
@@ -0,0 +1,39 @@
+	normalizeForDisplay,
+} from 'state/posts/revisions/selectors';
+
+class EditorDiffViewer extends PureComponent {
```
Why `PureComponent`? Are there any tradeoffs to consider? Would a simpler functional component work?

bperson replied:
For now a functional component would work, I tend to start with components while building as it allows to quickly add methods and lifecycle-related stuff without updating a lot of lines / re-indenting stuff around (now I wonder if there isn't an atom or emacs plugin that would allow me to switch between a component form and a functional one in a few keystrokes).

mtias commented on client/post-editor/editor-diff-viewer/index.jsx:
```diff
@@ -0,0 +1,39 @@
+	render() {
+		return (
+			<div className="editor-diff-viewer">
+				<h3 className="editor-diff-viewer__title">
```
Curious why did you choose an `h3` here.

bperson replied:
No real reasoning :) Updated to a `h1` since it can be considered as the "main" title of the page when the diff viewer is shown?

mtias commented on client/post-editor/editor-diff-viewer/index.jsx:
```diff
@@ -0,0 +1,39 @@
+
+export default connect(
+	( state, ownProps ) => ( {
+		revision: normalizeForDisplay( getPostRevision( state, ownProps.siteId, ownProps.postId, ownProps.selectedRevisionId ) ),
```
This line looks a bit funny with the nested calls and long prop attributes. Could it be simplified with destructuring?

bperson replied:
Updated to:
```
revision: normalizeForDisplay(
    getPostRevision( state, ownProps.siteId, ownProps.postId, ownProps.selectedRevisionId )
),
```
which makes it consistent with the indentation used in `editor-revisions-list/index.jsx`.

Other solutions:
- `flow`-ing for 2 functions (and one with so many parameters) is probably overkill
- destructing AFAIK would mean updating `getPostRevision` to: `getPostRevision = ( state, { siteId, postId, revisionId} )` which means we can't use `createSelector` anymore because the second arg is effectively an object and we only accept `number`, `boolean` or `string` [0]

[0] https://github.com/Automattic/wp-calypso/blob/master/client/lib/create-selector/index.js#L17

mtias commented on client/post-editor/editor-diff-viewer/style.scss:
```diff
@@ -0,0 +1,20 @@
+	color: $gray-dark;
+	font-weight: 600;
+	resize: none;
+	-ms-overflow-y: hidden !important;
```
Why is this needed?

bperson replied:
copy-pasted from the the editor's title input :) some (all? :D ) are definitely useless.

bperson replied:
kept `color` and `font-weight` in the latest update, the rest is out :)

mtias commented on client/post-editor/editor-drawer/index.jsx:
```diff
@@ -85,6 +85,7 @@ const EditorDrawer = React.createClass( {
 		type: React.PropTypes.string,
 		setPostDate: React.PropTypes.func,
 		onSave: React.PropTypes.func,
+		toggleChildSidebar: React.PropTypes.func,
```
Same comment as above, could it be more descriptive?

mtias commented on client/post-editor/editor-revisions-list/header.jsx:
```diff
@@ -0,0 +1,36 @@
+		return (
+			<div className="editor-revisions-list__header">
+				<h3>
+					{ this.props.translate( 'You\'re seeing a past revision' ) }
```
Instead of escaping could use double quotes.

bperson replied:
fixed

mtias commented on client/post-editor/editor-revisions-list/index.jsx:
```diff
@@ -0,0 +1,93 @@
+				<ul className="editor-revisions-list__list">
+					{ map( this.props.revisions, revision => (
+						<li
+							className={ classNames(
```
Doing the classnames construct outside the return call often helps with readability.

bperson replied:
fixed

mtias commented on client/post-editor/editor-revisions-list/style.scss:
```diff
@@ -0,0 +1,59 @@
+	&.selected {
+		background: $blue-medium;
+
+		&, *, .button {
```
What's the purpose of such a generic selector?

bperson replied:
To "force" a white color with a quite heavy specificity, there are a couple of ways out of "*", none are worth it imho:
- `&, span, .button` -> `span` is (almost) as generic and carries a toned down message than: "let's white all the thingz!" which is really what we want here :) 
- `&, .editor-revisions-list__deletions, .editor-revisions-list__additions, .button, ... ` (with `...` containing a lot of classes) this will constantly break each time the markup is updated.
- `&, .button` + a few blocks like this:

```
.editor-revision-list__deletions {
    .is-selected & {
        color: $white
    }
}
```
This introduces considerably more code and makes it harder to change the `is-selected` style for everything (though we can minimize that if we anticipate to only change the color with: `$revision-selected-color: $white`)

mtias commented on client/post-editor/editor-revisions-list/index.jsx:
```diff
@@ -0,0 +1,93 @@
+						<li
+							className={ classNames(
+								'editor-revisions-list__revision',
+								{ selected: revision.id === this.props.selectedRevisionId }
```
Take a look at the [CSS guidelines](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md), classes like these should be written as `is-selected`.

bperson replied:
fixed

mtias commented on client/post-editor/editor-revisions/index.jsx:
```diff
@@ -1,53 +1,45 @@
-} );
+
+	return (
+		<button
```
Why is this a `button` instead of `Button`?

bperson replied:
Removed the `a` because it's not a link anymore, went for a `button` to not load all the `Button` related CSS that would imply creating even more changes to the scss. I tried to minimise the amount of disruption there since it's a part that will be retouched when implementing @folletto status accordion redesign in a future PR.

mtias commented on client/post-editor/editor-sidebar/index.jsx:
```diff
@@ -22,40 +26,96 @@ export default class EditorSidebar extends Component {
+
+		const sidebarClassNames = classNames(
+			'post-editor__sidebar',
+			{ 'focus-child': this.state.childSidebar !== CHILD_SIDEBAR_NONE }
```
Same case with this class name.

mtias commented on client/post-editor/editor-sidebar/index.jsx:
```diff
@@ -22,40 +26,96 @@ export default class EditorSidebar extends Component {
-				/>
-				<SidebarFooter>
-					<EditorDeletePost
+				<div className="editor-sidebar__parent">
```
Could this have more descriptive names instead of parent/child.

mtias commented:
How would you go about making smaller PRs for these changes so it's easier to review and test?

bperson commented:
> How would you go about making smaller PRs for these changes so it's easier to review and test?

Probably something along those lines:
1. `requestUsers` actions + related data-layer files
2. updates to the post revisions selectors + updates to the revision normalisation in the data layer + the `requestUsers` trigger once we've received a couple of revisions.
3. introduction of the "child" sidebar concept (I'll try to figure out a better "naming" than child / parent, but until now, I couldn't find one: `subSidebar`, `detailsSidebar` aren't really more expressive)
4. post revisions list
5. diff viewer
6. editor wiring to load a revision

But before splitting, I would like to have a working prototype that computes the diffs between revisions, to see how it will fit in all that ;) 

Thanks for the review @mtias

bperson commented:
For the child/parent sidebar naming, I searched a bit more to find a more descriptive name for this kind of construct (in order, from more to less descriptive imho):
- `nestedSidebar`
- `cascadingSidebar`: inspired by the usage of cascading menus, but haven't really seen any usage of it with sidebars.
- `subSidebar`
- `childSidebar`
- `detailsSidebar`
- `secondarySidebar`

@mtias What do you think?

bperson commented:
> If you notice, the footer is meant to be present, but without "Move to trash", so that should answer the question. :D

@folletto hehe nice pre-answer :) it's been updated, current result:

![footer](https://cloud.githubusercontent.com/assets/1145270/25652838/5e20597a-2fea-11e7-9645-8d0aa1db84d7.png)

That said, I wonder if the footer and the block to "load revision in the editor" should be fixed? Right now, to load a revision from a while back, you have to scroll down, select the revision, scroll back up to load the revision in the editor.

When we have a lot of revisions, the footer is hardly reachable right now. And it may get even more unusable if we implement lazy loading / infinite scrolling since it will get bumped around.

Is there any issue with fixing those 2 blocks?

folletto commented:
>  I wonder if the footer and the block to "load revision in the editor" should be fixed? 

That's a fair question.

* Footer — I think we just missed making it sticky. Notice however that it should _still_ scroll on mobile sizes to optimize space. Check how the My Sites sidebar behaves there to mirror the solution.
* Load Revision — Good point. I'm not sure, so let's try to stick it, and see how it feels. :)

bperson commented:
### Update
- the sidebar's header and footer are now fixed on desktop, scroll on mobiles same as the "my sites" sidebar. (I still need to test "fixing" the "load the revision" block though)
- add dependency on https://github.com/kpdecker/jsdiff
- compute the diff between revisions:
    - when loading the revisions from the API, saving the changes summary ( words added / words removed ) directly in Redux
    - when displaying a revision in the diff viewer, it re-computes the complete diff to display all changes, memoized via the `getPostRevisionChanges` selector.

<img width="1425" alt="screen shot 2017-05-04 at 22 16 32" src="https://cloud.githubusercontent.com/assets/1145270/25723107/f9f92da4-3116-11e7-946b-c53730a91c93.png">

bperson commented:
### Update
- if a child sidebar is active ( revisions for instance ), the editor's ground control icon for the sidebar toggle is updated.
- mobile / smaller screens:
    - the sidebar's headers is displayed for "children" sidebars (this provides a way of dismissing the child sidebar)
    - do not pre-select a revision when displaying all revisions
    - on revision selection, the layout focus is switched to the content to "see" the changes introduced by that revision, you then have to open again the sidebar to "load the revision" in the editor.
    - when loading a revision in the editor, the layout focus is set to content to dismiss the sidebar.
- transform all eligible components to functional components

The last couple of issues (scrolling, placeholders when loading revisions, ... ) can be dealt with after the big split&trade; into smaller PRs.

bperson commented:
![mobile-flow](https://cloud.githubusercontent.com/assets/1145270/25848293/036c7660-34ba-11e7-80ed-aef118688877.gif)

@folletto After having played with the mobile version a bit more: we need to open the sidebar again to load the revision in the editor. Is there a reason for not placing that "load in the editor" in the main content view? ( right of the title or maybe replacing the "publish/update" button when in revision mode?). This would also solve the problem mentioned earlier on this PR where I talked about "fixing" the block to constantly have access to it.

Or we can do that in a subsequent PR :) I'm thinking about starting The big split&trade; into smaller PRs tonight / tomorrow.

folletto commented:
> Is there a reason for not placing that "load in the editor" in the main content view?

The reason is mainly to start having a v1 that doesn't make things too complicated on mobile as having to move around components across body and sidebar.

>  ( right of the title or maybe replacing the "publish/update" button when in revision mode?).

It's important to not change the top navigation and publish/update because it's a reminder that the change isn't going to go live. We should try to be very clear to the users about this, and avoid the confusion "Is it going to be posted?".

bperson commented:
> The reason is mainly to start having a v1 that doesn't make things too complicated on mobile as having to move around components across body and sidebar.

deal :D 

I started The Big Split™, Github will auto-reference all smaller PRs here which should help overall tracking.

mcsf commented:
Closing in favor of #14242.
